exercises/tom_hampson/ex02_oo_basics/src/ShapeSorter.cpp

```diff
@@ -0,0 +1,126 @@
+#include "ShapeSorter.h"
+
+#include <algorithm>
+#include <iostream>
+
+ShapeSorter::~ShapeSorter()
+{
+}
+
+/**
+ * @brief Print all shapes with specified shape type (e.g. "circle").
+ * @param shapeType 
+*/
+void ShapeSorter::printShapesOfType(const std::string& shapeType) const
+{
+  std::cout << "Shapes of type \"" << shapeType << "\":" << std::endl;
+
+  for (auto shape : m_shapes)
+  {
+    if (shape->shapeType() == shapeType)
+    {
+      shape->print();
+    }
+  }
+}
+
+/**
+ * @brief Print all shapes with n sides.
+ * @param n 
+*/
+void ShapeSorter::printShapesWithNSides(const int n) const
+{
+  std::cout << "Shapes with " << n << " sides:" << std::endl;
+  for (auto shape : m_shapes)
+  {
+    if (shape->nSides() == n)
+    {
+      shape->print();
+    }
+  }
+}
+
+/**
+ * @brief Print all shapes by area in descending order.
+*/
+void ShapeSorter::printShapesAreaDescending() const
+{
+  //Make a copy of the shapes vector so we preserve the order of the member shape vector.
+  std::vector<std::shared_ptr<Shape>> sortedShapes = m_shapes;
+  std::sort(sortedShapes.begin(), sortedShapes.end(), compareArea);
+
+  std::cout << "Shapes ordered by area:" << std::endl;
+  for (auto shape : sortedShapes)
+  {
+    shape->print();
+  }
+}
+
+/**
+ * @brief Print all shapes in descending order of perimeter size.
+*/
+void ShapeSorter::printShapesPerimeterDescending() const
+{
+  //Make a copy of the shapes vector so we preserve the order of the member shape vector.
+  std::vector<std::shared_ptr<Shape>> sortedShapes = m_shapes;
+  std::sort(sortedShapes.begin(), sortedShapes.end(), comparePerimeter);
+
+  std::cout << "Shapes ordered by perimeter:" << std::endl;
+  for (auto shape : sortedShapes)
```

Comment:
Again could use const reference in these loops.

Reply:
done